### PR TITLE
[AMDGPU] Respect existing glue when lowering convergence tokens

### PR DESCRIPTION
**Pull request by ssahasra**

### CODE REVIEW

**Comment by jayfoad** [comment]
I guess no nodes actually require this yet. Are you expecting any? Or is this just defensive programming?

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -3859,20 +3859,20 @@ SDValue SITargetLowering::LowerCall(CallLoweringInfo &CLI,
...
-  // Emitter::EmitMachineNode() expects only the glued convergence token if it
-  // exists. Similarly, the selection of the call expects to match only the
-  // InGlue operand if it exists.
   if (SDValue Token = CLI.ConvergenceControlToken) {
-    Ops.push_back(SDValue(DAG.getMachineNode(TargetOpcode::CONVERGENCECTRL_GLUE,
-                                             DL, MVT::Glue, Token),
-                          0));
+    SmallVector<SDValue, 2> GlueOps;
+    GlueOps.push_back(Token);
+    if (InGlue.getNode())
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:3865]: Don't need .getNode()

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 2]
```diff
@@ -3859,20 +3859,20 @@ SDValue SITargetLowering::LowerCall(CallLoweringInfo &CLI,
...
+      GlueOps.push_back(InGlue);
+
+    InGlue = SDValue(DAG.getMachineNode(TargetOpcode::CONVERGENCECTRL_GLUE, DL,
+                                        MVT::Glue, GlueOps),
+                     0);
   }
 
+  if (InGlue.getNode())
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:3873]: Don't need .getNode()

**Comment by ssahasra** [comment]
> I guess no nodes actually require this yet. Are you expecting any? Or is this just defensive programming?

What is "this" in this context?? When lowering calls, we do glue stuff onto the actually call instruction. Now we also need to glue the convergence token to the same call. The existing attempt happily created a second glue operand, and this change fixes that to have a single glue operand which could itself have a chain of glued operands.

**Comment by jayfoad** [comment]
> > I guess no nodes actually require this yet. Are you expecting any? Or is this just defensive programming?
> 
> What is "this" in this context??

"This" meant "this patch". I was trying to work out what kind of patch this is. Does it fix a bug? Why is there no test case?

**Comment by ssahasra** [comment]
> > > I guess no nodes actually require this yet. Are you expecting any? Or is this just defensive programming?
> 
> "This" meant "this patch". I was trying to work out what kind of patch this is. Does it fix a bug? Why is there no test case?

This patch merely cements my improved understanding of how glue is supposed to be set on an operation. Callsites already have a glue operand that holds on to the arg nodes. The lowering of control flow tokens at a callsite was adding a second glue operand, which is unexpected. This patch cleans that up by chaining the token with the existing glue operand. In this narrow window inside the AMDGPU backend, this "bug" would have been triggered if those arg operands were actually accessed as glue. But they are not, so nothing broke when I put a second glue operand.

**Pull request merged**